geopandas/io/sql.py

```diff
@@ -98,3 +101,257 @@ def load_geom_text(x):
                 crs = "epsg:{}".format(srid)
 
     return GeoDataFrame(df, crs=crs, geometry=geom_col)
+
+
+def _get_geometry_type(gdf):
+    """
+    Get basic geometry type of a GeoDataFrame. See more info from:
+    https://geoalchemy-2.readthedocs.io/en/latest/types.html#geoalchemy2.types._GISType
+
+    Following rules apply:
+     - if geometries all share the same geometry-type,
+       geometries are inserted with the given GeometryType with following types:
+        - Point, LineString, Polygon, MultiPoint, MultiLineString, MultiPolygon,
+          GeometryCollection.
+        - LinearRing geometries will be converted into LineString -objects.
+     - in all other cases, geometries will be inserted with type GEOMETRY:
+        - a mix of Polygons and MultiPolygons in GeoSeries
+        - a mix of Points and LineStrings in GeoSeries
+        - geometry is of type GeometryCollection,
+          such as GeometryCollection([Point, LineStrings])
+     - if any of the geometries has Z-coordinate, all records will
+       be written with dimension=3.
+     """
+    geom_types = list(gdf.geometry.geom_type.unique())
+    has_curve = False
+
+    # Check for 3D-coordinates
+    dimension = 2
+    management = False
+    if any(gdf.geometry.has_z):
+        dimension = 3
+        management = True
```

Comment:
I was looking at the geoalchemy2 docs (https://github.com/geoalchemy/geoalchemy2/blob/710719827e77a8c186ba863a0b200b869b4079df/geoalchemy2/types.py#L89-L91):

> With ``management`` set to ``False`` (the default) ``dimension`` is not
        taken into account, and the actual dimension is fully defined with the
        ``geometry_type``.

and it seems that they then add "Z" to the geometry type instead of using `dimension=3, management=True` (so using "GEOMETRYZ", "POINTZ", ..)

That might be the better option to avoid needing to do a "AddGeometryColumn" (which seems is not used anymore by default / recommended in postgis >= 2)

Reply:
So I think adding something like 

```
    if any(gdf.geometry.has_z):
        target_geom_type = target_geom_type + "Z"
```

should also do the trick

Reply:
Ah, great! Let's do it that way :+1: